src/topology/metric_space/contracting.lean

```diff
@@ -39,17 +46,164 @@ def contracting_with [emetric_space α] (K : ℝ≥0) (f : α → α) :=
 
 namespace contracting_with
 
-variables [metric_space α] {K : ℝ≥0} {f : α → α} (hf : contracting_with K f)
-
+variables [emetric_space α] [cs : complete_space α] {K : ℝ≥0} {f : α → α}
+  (hf : contracting_with K f)
 include hf
 
+open emetric set
+
 lemma to_lipschitz_with : lipschitz_with K f := hf.2
 
+lemma one_sub_K_pos' : (0:ennreal) < 1 - K := by simp [hf.1]
+
+lemma one_sub_K_ne_zero : (1:ennreal) - K ≠ 0 := ne_of_gt hf.one_sub_K_pos'
+
+lemma one_sub_K_ne_top : (1:ennreal) - K ≠ ⊤ :=
+by { norm_cast, exact ennreal.coe_ne_top }
+
+lemma edist_inequality (hf : contracting_with K f) {x y} (h : edist x y < ⊤) :
+  edist x y ≤ (edist x (f x) + edist y (f y)) / (1 - K) :=
+suffices edist x y ≤ edist x (f x) + edist y (f y) + K * edist x y,
+  by rwa [ennreal.le_div_iff_mul_le hf.one_sub_K_ne_zero hf.one_sub_K_ne_top,
+    mul_comm, ennreal.sub_mul (ne_of_lt h), one_mul, ennreal.sub_le_iff_le_add],
+calc edist x y ≤ edist x (f x) + edist (f x) (f y) + edist (f y) y : edist_triangle4 _ _ _ _
+  ... = edist x (f x) + edist y (f y) + edist (f x) (f y) : by rw [edist_comm y, add_right_comm]
+  ... ≤ edist x (f x) + edist y (f y) + K * edist x y : add_le_add' (le_refl _) (hf.2 _ _)
+
+lemma edist_le_of_fixed_point {x y} (h : edist x y < ⊤) (hy : f y = y) :
+  edist x y ≤ (edist x (f x)) / (1 - K) :=
+by simpa only [hy, edist_self, add_zero] using hf.edist_inequality h
+
+lemma eq_or_edist_eq_top_of_fixed_points {x y} (hx : f x = x) (hy : f y = y) :
+  x = y ∨ edist x y = ⊤ :=
+begin
+  cases eq_or_lt_of_le (lattice.le_top : edist x y ≤ ⊤), from or.inr h,
+  refine or.inl (edist_le_zero.1 _),
+  simpa only [hx, edist_self, add_zero, ennreal.zero_div]
+    using hf.edist_le_of_fixed_point h hy
+end
+
+/-- Banach fixed-point theorem, contraction mapping theorem, `emetric_space` version.
+This version assumes that `f` is contracting in the whole space and sends a point `x`
+of a forward-invariant complete set `s` to the same ball of radius infinity.
```

Comment:
This seems to me to be a weird set of assumptions: the most usual situation is when `f` is contracting on the whole space, and the most general one is when `f` is contracting on a forward invariant set. You are somewhat inbetween, assuming contraction on the whole space but arguing only in a forward invariant set. Do you have a motivating case for this intermediate degree of generality?

Reply:
I had a motivation but now I don't think it was valid. I need this theorem for a function contracting on a non-complete space covered by forward-invariant complete sets. In particular, I want to have uniqueness in this case.
I think that I'll prove this theorem (and define `efixed_point`) for a function contracting on a forward-invariant subset, then prove a version of uniqueness lemma for two forward-invariant sets.

P.S.: I'd prefer to prove it in a complete space, then apply to `λ x : s, ⟨f x, hfs x.2⟩`, where `hfs : maps_to f s s`, but conversion to/from subtype is not automatic.

Reply:
Done.